Mathlib/Algebra/Lie/CartanExists.lean

```diff
@@ -0,0 +1,392 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import Mathlib.Algebra.Lie.EngelSubalgebra
+import Mathlib.Algebra.Lie.CartanSubalgebra
+import Mathlib.Algebra.Lie.Rank
+import Mathlib.LinearAlgebra.Eigenspace.Zero
+
+/-!
+# Existence of Cartan subalgebras
+
+In this file we prove existence of Cartan subalgebras in finite-dimensional Lie algebras,
+following [barnes1967].
+
+## Main results
+
+* `exists_IsCartanSubalgebra_of_finrank_le_card`:
+  A Lie algebra `L` over a field `K` has a Cartan subalgebra,
+  provided that the dimension of `L` over `K` is less than or equal to the cardinality of `K`.
+* `exists_IsCartanSubalgebra`:
+  A finite-dimensional Lie algebra `L` over an infinite field `K` has a Cartan subalgebra.
+
+## References
+
+* [barnes1967]: "On Cartan subalgebras of Lie algebras" by D.W. Barnes.
+
+-/
+
+-- move this to `Mathlib.Algebra.MvPolynomial.Basic`
+lemma MvPolynomial.coe_aeval_eq_eval {R σ : Type*} [CommRing R] (x : σ → R) :
+    RingHomClass.toRingHom (MvPolynomial.aeval x) = MvPolynomial.eval x := rfl
+
+-- move this to `Mathlib.LinearAlgebra.FiniteDimensional`
+open FiniteDimensional in
+lemma _root_.Submodule.eq_top_of_finrank
+    {K V : Type*} [DivisionRing K] [AddCommGroup V] [Module K V] [FiniteDimensional K V]
+    (W : Submodule K V) (h : finrank K W = finrank K V) :
+    W = ⊤ :=
+  eq_of_le_of_finrank_le le_top <| by rw [finrank_top, h]
+
+namespace LieAlgebra
+
+section CommRing
+
+variable {K R L M : Type*}
+variable [Field K] [CommRing R] [Nontrivial R]
+variable [LieRing L] [LieAlgebra K L] [LieAlgebra R L]
+variable [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+variable [Module.Finite K L]
+variable [Module.Finite R L] [Module.Free R L]
+variable [Module.Finite R M] [Module.Free R M]
+
+open FiniteDimensional LieSubalgebra Module.Free Polynomial
+
+variable (K)
+
+lemma finrank_engel (x : L) :
+    finrank K (engel K x) = (ad K L x).charpoly.natTrailingDegree :=
+  (ad K L x).finrank_maximalGeneralizedEigenspace
+
+lemma rank_le_finrank_engel (x : L) :
+    rank K L ≤ finrank K (engel K x) :=
+  (rank_le_natTrailingDegree_charpoly_ad K x).trans
+    (finrank_engel K x).ge
+
+lemma isRegular_iff_finrank_engel_eq_rank (x : L) :
+    IsRegular K x ↔ finrank K (engel K x) = rank K L := by
+  rw [isRegular_iff_natTrailingDegree_charpoly_eq_rank, finrank_engel]
+
+variable {K}
```

Comment:
I think you can move this line to line 92 with the same effect.

Reply:
In fact, I could just remove it altogether.